### PR TITLE
Demonstrate converter for Dictionary<TKey,> where TKey is not String

### PR DESCRIPTION
**Pull request by steveharter**
Based on questions from https://github.com/dotnet/corefx/issues/40120 and elsewhere, demonstrate how a custom converter can be written to handle this for the 3.x timeframe.

In 5.0 we may have a way to do add a built-in converter pending semantics.

These tests may be ported to release/3.0 in order to prevent regressions.

### CODE REVIEW

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryConverterForIDictionary.cs [hunk 1]
```diff
@@ -9,17 +9,25 @@ namespace System.Text.Json.Serialization.Tests
 {
     public static partial class CustomConverterTests
     {
-        // Demonstrates custom Dictionary<string, long>; Adds offset to each integer or long to verify converter ran.
-        internal class DictionaryConverter : JsonConverter<Dictionary<string, long>>
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryConverterForIDictionary.cs]: Are we losing some test case coverage by not having a concrete `Dictionary<TKey, TValue>` based converter, or do you feel confident the scenario is covered by the other tests/changes here?

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryGuidConverter.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+using System.Reflection;
+using Xunit;
+
+namespace System.Text.Json.Serialization.Tests
+{
+    public static partial class CustomConverterTests
+    {
+        /// <summary>
+        /// Demonstrates a <see cref="Dictionary{Guid, TValue}"> converter using a JSON object with property names representing keys.
+        /// Sample JSON for <see cref="Dictionary{Guid, object}">: {"2E6E1787-1874-49BF-91F1-0F65CCB6C161":{}}
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryGuidConverter.cs]: Should this be TValue instead of object or am I misunderstanding the intent here?
```suggestion
        /// Sample JSON for <see cref="Dictionary{Guid, TValue}">: {"2E6E1787-1874-49BF-91F1-0F65CCB6C161":{}}
```

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryGuidConverter.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+                        if (!Guid.TryParseExact(propertyName, format:"D", out Guid result))
+                        {
+                            throw new JsonException($"Unable to convert \"{propertyName}\" to a Guid.");
+                        }
+
+                        // Get the value.
+                        TValue v;
+                        if (_valueConverter != null)
+                        {
+                            reader.Read();
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryGuidConverter.cs]: Interesting that the reader needs to be set to past the value before passing it into the `_valueConverter`. Looking at this from outside in, I would have expected the converter to do a single read itself.

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryGuidConverter.cs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+                    culture: null);
+
+                return converter;
+            }
+
+            private class DictionaryGuidConverterInner<TValue> : JsonConverter<Dictionary<Guid, TValue>>
+            {
+                private readonly JsonConverter<TValue> _valueConverter;
+
+                public DictionaryGuidConverterInner(JsonSerializerOptions options)
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryGuidConverter.cs]: null check on options?

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryKeyValueConverter.cs [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+                return converter;
+            }
+
+            private class DictionaryKeyValueConverterInner<TKey, TValue> : JsonConverter<Dictionary<TKey, TValue>>
+            {
+                private readonly JsonConverter<KeyValuePair<TKey, TValue>> _converter;
+
+                public DictionaryKeyValueConverterInner(JsonSerializerOptions options)
+                {
+                    _converter = (JsonConverter<KeyValuePair<TKey, TValue>>)options.GetConverter(typeof(KeyValuePair<TKey, TValue>));
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryKeyValueConverter.cs]: Same here, null check?

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryKeyValueConverter.cs [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+            public Dictionary<MyEnum, Entity> EnumKey { get; set; }
+            public Dictionary<Guid, Entity> GuidKey { get; set; }
+            public Dictionary<int, Entity> IntKey { get; set; }
+            public Dictionary<float, Entity> FloatKey { get; set; }
+            public Dictionary<double, Entity> DoubleKey { get; set; }
+            public Dictionary<string, Entity> StringKey { get; set; }
+        }
+
+        [Fact]
+        public static void AllPrimitivesConvertion()
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryKeyValueConverter.cs]:
```suggestion
        public static void AllPrimitivesConversion()
```

File: src/System.Text.Json/tests/System.Text.Json.Tests.csproj [hunk 7]
```diff
@@ -129,4 +132,7 @@
     <Compile Include="JsonObjectTests.cs" />
     <Compile Include="JsonStringTests.cs" />
   </ItemGroup>
+  <ItemGroup>
+    <None Include="Serialization\CustomConverterTests.DictionaryInt32StringKeyValueConverter.cs" />
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/System.Text.Json.Tests.csproj]: Why is this file added as a separate group with "None"?

File: src/System.Text.Json/tests/System.Text.Json.Tests.csproj [hunk 8]
```diff
@@ -40,7 +40,10 @@
     <Compile Include="Serialization\CustomConverterTests.BadConverters.cs" />
     <Compile Include="Serialization\CustomConverterTests.cs" />
     <Compile Include="Serialization\CustomConverterTests.DerivedTypes.cs" />
-    <Compile Include="Serialization\CustomConverterTests.Dictionary.cs" />
+    <Compile Include="Serialization\CustomConverterTests.DictionaryConverterForIDictionary.cs" />
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/System.Text.Json.Tests.csproj]: nit: Be consistent in file names.

`CustomConverterTests.IDictionaryConverter`?

It looks like the other converter files have `Converter` as a suffix.

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryConverterForIDictionary.cs [hunk 9]
```diff
@@ -9,17 +9,25 @@ namespace System.Text.Json.Serialization.Tests
...
-            public DictionaryConverter(long offset)
+            public DictionaryConverterForIDictionary(long offset)
             {
                 _offset = offset;
             }
 
-            public override Dictionary<string, long> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            public override bool CanConvert(Type typeToConvert)
+            {
+                return typeToConvert == typeof(Dictionary<string, long>);
```
[the remark this answers is not in the document]
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryConverterForIDictionary.cs]: It is OK to say the converter is for `IDictionary` when only `Dictionary` is supported -- that's one reason why `CanConvert` exists. The sample could be extended to support other dictionary types.

Given that, everything could just be `Dictionary` however for sample purposes.

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryConverterForIDictionary.cs (hunk 1 above)
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryConverterForIDictionary.cs]: This example is for a simple Dictionary converter for specific types. The other samples provide more complicated scenarios that use the factory pattern and can support any TKey and\or TValue.

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryGuidConverter.cs (hunk 2 above)
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryGuidConverter.cs]: Since this supports any TValue, the sample is just for an object which has {}. A sample for an integer would not have {} for example.

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryGuidConverter.cs (hunk 4 above)
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryGuidConverter.cs]: Since it is a private nested class, it is not possible to pass in a null.

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryKeyValueConverter.cs (hunk 6 above)
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryKeyValueConverter.cs]: Thanks will update.

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryConverterForIDictionary.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryConverterForIDictionary.cs]: I'll rename the test and make sure it's not considered an example. This was intended for increasing code coverage for the "contravariant" code in JsonPropertyInfo.

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryConverterForIDictionary.cs (hunk 1 above)
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryConverterForIDictionary.cs]: This was an existing test - just renamed. The new samples provide for any TKey and\or any TValue.

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryKeyValueConverter.cs (hunk 5 above)
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/CustomConverterTests.DictionaryKeyValueConverter.cs]: Same as above - it is not possible for the options to be null.

File: src/System.Text.Json/tests/System.Text.Json.Tests.csproj (hunk 7 above)
Comment (steveharter) [reply src/System.Text.Json/tests/System.Text.Json.Tests.csproj]: Good catch. Not sure but will fix.

File: src/System.Text.Json/tests/System.Text.Json.Tests.csproj (hunk 8 above)
Comment (steveharter) [reply src/System.Text.Json/tests/System.Text.Json.Tests.csproj]: Per above comment -- this was an existing test -- changing name to ContravariantDictionaryConverter.cs
